sklearn/model_selection/tests/test_successive_halving.py

```diff
@@ -0,0 +1,403 @@
+import pytest
+from scipy.stats import norm, randint
+import numpy as np
+
+from sklearn.datasets import make_classification
+from sklearn.dummy import DummyClassifier
+from sklearn.model_selection import HalvingGridSearchCV
+from sklearn.model_selection import HalvingRandomSearchCV
+from sklearn.model_selection import KFold, ShuffleSplit
+from sklearn.model_selection._search_successive_halving import (
+    _SubsampleMetaSplitter, _top_k, _refit_callable)
+
+
+class FastClassifier(DummyClassifier):
+    """Dummy classifier that accepts parameters a, b, ... z.
+
+    These parameter don't affect the predictions and are useful for fast
+    grid searching."""
+
+    def __init__(self, strategy='stratified', random_state=None,
+                 constant=None, **kwargs):
+        super().__init__(strategy=strategy, random_state=random_state,
+                         constant=constant)
+
+    def get_params(self, deep=False):
+        params = super().get_params(deep=deep)
+        for char in range(ord('a'), ord('z') + 1):
+            params[chr(char)] = 'whatever'
+        return params
+
+
+@pytest.mark.parametrize('Est', (HalvingGridSearchCV, HalvingRandomSearchCV))
+@pytest.mark.parametrize(
+    ('aggressive_elimination,'
+     'max_resources,'
+     'expected_n_iterations,'
+     'expected_n_required_iterations,'
+     'expected_n_possible_iterations,'
+     'expected_n_remaining_candidates,'
+     'expected_n_resources,'), [
+         # notice how it loops at the beginning
+         (True, 'limited', 4, 4, 3, 1, [20, 20, 60, 180]),
+         # no aggressive elimination: we end up with less iterations and more
+         # candidates at the end
+         (False, 'limited', 3, 4, 3, 3, [20, 60, 180]),
+         # When the amount of resource isn't limited, aggressive_elimination
+         # has no effect. Here the default min_resources='exhaust' will take
+         # over.
+         (True, 'unlimited', 4, 4, 4, 1, [37, 111, 333, 999]),
+         (False, 'unlimited', 4, 4, 4, 1, [37, 111, 333, 999]),
+     ]
+)
+def test_aggressive_elimination(
+        Est, aggressive_elimination, max_resources, expected_n_iterations,
+        expected_n_required_iterations, expected_n_possible_iterations,
+        expected_n_remaining_candidates, expected_n_resources):
+    # Test the aggressive_elimination parameter.
+
+    n_samples = 1000
+    X, y = make_classification(n_samples=n_samples, random_state=0)
+    parameters = {'a': ('l1', 'l2'), 'b': list(range(30))}
+    base_estimator = FastClassifier()
+
+    if max_resources == 'limited':
+        max_resources = 180
+    else:
+        max_resources = n_samples
+
+    sh = Est(base_estimator, parameters,
+             aggressive_elimination=aggressive_elimination,
+             max_resources=max_resources, ratio=3,
+             verbose=True)  # just for test coverage
```

Comment:
might be best to put the comment on a line just with verbose=True. This comment will be confusing after we apply black, for instnace.